Source/OIDAuthorizationService.m

```diff
@@ -348,6 +358,120 @@ + (void)performTokenRequest:(OIDTokenRequest *)request callback:(OIDTokenCallbac
+        dispatch_async(dispatch_get_main_queue(), ^{
+          callback(nil, invalidIDToken);
+        });
+        return;
+      }
+
+      // OpenID Connect Core Section 3.1.3.7. rule #3
+      // Validates that the audience of the ID Token matches the client ID.
+      NSString *clientID = tokenResponse.request.clientID;
+      if (![idToken.audience containsObject:clientID]) {
+        NSError *invalidIDToken =
+          [OIDErrorUtilities errorWithCode:OIDErrorCodeIDTokenFailedValidationError
+                           underlyingError:nil
+                               description:@"Audience mismatch"];
+        dispatch_async(dispatch_get_main_queue(), ^{
+          callback(nil, invalidIDToken);
+        });
+        return;
+      }
+      
+      // OpenID Connect Core Section 3.1.3.7. rules #4 & #5
+      // Not supported.
+
+      // OpenID Connect Core Section 3.1.3.7. rule #6
+      // As noted above, AppAuth only supports the code flow which results in direct communication
+      // of the ID Token from the Token Endpoint to the Client, and we are exercising the option to
+      // use TSL server validation instead of checking the token signature. Users may additionally
+      // check the token signature should they wish.
+
+      // OpenID Connect Core Section 3.1.3.7. rules #7 & #8
+      // Not applicable. See rule #6.
+
+      // OpenID Connect Core Section 3.1.3.7. rule #9
+      // Validates that the current time is before the expiry time.
+      NSTimeInterval expiresAtDifference = [idToken.expiresAt timeIntervalSinceNow];
+      if (expiresAtDifference < 0) {
+        NSError *invalidIDToken =
+        [OIDErrorUtilities errorWithCode:OIDErrorCodeIDTokenFailedValidationError
+                         underlyingError:nil
+                             description:@"ID Token expired"];
+        dispatch_async(dispatch_get_main_queue(), ^{
+          callback(nil, invalidIDToken);
+        });
+        return;
+      }
+      
+      // OpenID Connect Core Section 3.1.3.7. rule #10
+      // Validates that the issued at time is not more than +/- 5 minutes on the current time.
+      NSTimeInterval issuedAtDifference = [idToken.issuedAt timeIntervalSinceNow];
+      if (fabs(issuedAtDifference) > 300) {
+        NSError *invalidIDToken =
+          [OIDErrorUtilities errorWithCode:OIDErrorCodeIDTokenFailedValidationError
+                           underlyingError:nil
+                               description:@"Issued at time is more than 5 minutes before or after "
+                                            "the current time"];
+        dispatch_async(dispatch_get_main_queue(), ^{
+          callback(nil, invalidIDToken);
+        });
+        return;
+      }
+      
+      // OpenID Connect Core Section 3.1.3.7. rule #11
+      // Validates the nonce.
+      NSString *nonce = authorizationResponse.request.nonce;
+      if (nonce && ![idToken.nonce isEqual:nonce]) {
```

Comment:
Not suer if you are already aware, but there is a problem here. A nonce isn't sent with the call and then it expects one here.

Reply:
If the authorization request doesn't set nonce, then this check will be skipped (`if (nonce &&`… should achieve that). Are you sure there's a problem?

Reply:
I have a project using this library with an IdentityServer4 server. The problem is that `authorizationResponse.request.nonce` is a valid nonce and idToken.nonce is null.

Reply:
[OpenID Connect Core Section 2](http://openid.net/specs/openid-connect-core-1_0.html#IDToken) states:

> If present in the Authentication Request, Authorization Servers MUST include a nonce Claim in the ID Token with the Claim Value being the nonce value sent in the Authentication Request. 

I believe we are implementing this correctly in AppAuth , is there a bug in IdentityServer4?

Reply:
I have opened a bug with IdentityServer4 here: https://github.com/IdentityServer/IdentityServer4/issues/2180

Reply:
[Section 12.1](http://openid.net/specs/openid-connect-core-1_0.html#RefreshingAccessToken)
Does not state anything about needing a nonce to verify a Refresh Request. Not sure AppAuth is really doing the right thing here.

Reply:
It's a bit ambiguous, but yeah I don't think the intention was for nonce to be required in the ID Token for the token response during refreshes. I sent a [note](http://lists.openid.net/pipermail/openid-specs-ab/Week-of-Mon-20180416/006728.html) to the Connect mailing list regarding the possible ambiguity. Google doesn't return the nonce in refreshed ID Tokens either.

I've fixed this in the PR. Thanks very much for engaging and reporting this problem!